e2e_tests/tests/cluster/test_checkpoints.py

```diff
@@ -75,25 +76,40 @@ def run_gc_checkpoints_test(checkpoint_storage: Dict[str, str]) -> None:
     # Check that the actual checkpoint storage (for shared_fs) reflects the
     # deletions. We want to wait for the GC containers to exit, so check
     # repeatedly with a timeout.
+
     max_checks = 30
     for i in range(max_checks):
         time.sleep(1)
         try:
+            storage_states = []
             for config, checkpoints in all_checkpoints:
                 checkpoint_config = config["checkpoint_storage"]
 
                 storage_manager = storage.build(checkpoint_config, container_path=None)
+                storage_state = {}
                 for checkpoint in checkpoints:
                     storage_id = checkpoint["uuid"]
+                    # Trying again
                     if checkpoint["state"] == "COMPLETED":
                         with storage_manager.restore_path(storage_id):
```

Comment:
There's a second type of failure possible here, which is implicit in the code.

If the checkpoint was supposed to exist but doesn't, this line will raise a CheckpointNotFound.  You'll have to catch that error and add an entry into your lists of issues as well.

Otherwise, this looks pretty good.

Reply:
Got it. Yeah, this makes more sense after I looked at it today.